torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,272 @@
+from torch.optim.lr_scheduler import CosineAnnealingLR
+from torchvision.models._api import WeightsEnum
+
+from ..models import get_weight
+from . import utils
+
+try:
+    from torch.optim.lr_scheduler import LRScheduler
+except ImportError:
+    from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+
+
+# https://github.com/google-research/simclr/blob/master/data_util.py
+SIZE = 224
+KS = SIZE // 10 // 2 * 2 + 1
+AUG = K.AugmentationSequential(
+    K.RandomResizedCrop(size=(SIZE, SIZE), ratio=(0.75, 1.33)),
+    K.RandomHorizontalFlip(),
+    K.RandomVerticalFlip(),  # added
+    # Not appropriate for multispectral imagery, seasonal contrast used instead
+    # K.ColorJitter(brightness=0.8, contrast=0.8, saturation=0.8, hue=0.2, p=0.8)
+    # K.RandomGrayscale(p=0.2),
+    K.RandomGaussianBlur(kernel_size=(KS, KS), sigma=(0.1, 2)),
+    data_keys=["input"],
+)
+
+
+# Remove once https://github.com/lightly-ai/lightly/pull/1150 is merged and released
+class SimCLRProjectionHead(ProjectionHead):  # type: ignore[misc]
+    """SimCLR projection head."""
+
+    def __init__(
+        self,
+        input_dim: int = 2048,
+        hidden_dim: int = 2048,
+        output_dim: int = 2048,
+        num_layers: int = 3,
+    ) -> None:
+        """Initialize a new SimCLRProjectionHead instance.
+
+        Args:
+            input_dim: Number of input dimensions.
+            hidden_dim: Number of hidden dimensions.
+            output_dim: Number of output dimensions.
+            num_layers: Number of hidden layers.
+        """
+        layers: list[tuple[int, int, Optional[nn.Module], Optional[nn.Module]]] = []
+        layers.append((input_dim, hidden_dim, nn.BatchNorm1d(hidden_dim), nn.ReLU()))
+        for _ in range(2, num_layers):
+            layers.append(
+                (hidden_dim, hidden_dim, nn.BatchNorm1d(hidden_dim), nn.ReLU())
+            )
+        layers.append((hidden_dim, output_dim, nn.BatchNorm1d(output_dim), None))
+        super().__init__(layers)
+
+
+class SimCLRTask(LightningModule):  # type: ignore[misc]
+    """SimCLR: a simple framework for contrastive learning of visual representations.
+
+    Reference implementation:
+
+    * https://github.com/google-research/simclr
+
+    If you use this trainer in your research, please cite the following papers:
+
+    * v1: https://arxiv.org/abs/2002.05709
+    * v2: https://arxiv.org/abs/2006.10029
+
+    .. versionadded:: 0.5
+    """
+
+    def __init__(
+        self,
+        model: str = "resnet50",
+        weights: Optional[Union[WeightsEnum, str, bool]] = None,
+        in_channels: int = 3,
+        version: int = 2,
+        layers: int = 3,
+        hidden_dim: Optional[int] = None,
+        output_dim: Optional[int] = None,
+        lr: float = 4.8,
+        weight_decay: float = 1e-4,
+        temperature: float = 0.07,
+        memory_bank_size: int = 64000,
+        gather_distributed: bool = False,
+        augmentations: nn.Module = AUG,
```

Comment:
Not saying we shouldn't make this a parameter, but the docs for this look rough...